mlir/lib/Dialect/Linalg/IR/LinalgInterfaces.cpp

```diff
@@ -1172,19 +1240,26 @@ LogicalResult mlir::linalg::detail::verifyStructuredOpInterface(Operation *op) {
 
     // Domain must be consistent.
     unsigned numLoops = linalgOp.getNumLoops();
-    if (indexingMap.getNumDims() != numLoops)
+    if (!isUserDefinedSemantic && indexingMap.getNumDims() != numLoops)
       return op->emitOpError("expected indexing_map #")
              << opOperand.getOperandNumber() << " to have " << numLoops
              << " dim(s) to match the number of loops";
 
     int64_t rank = linalgOp.getRank(&opOperand);
-    if (indexingMap.getNumResults() != rank)
+
+    if (!isUserDefinedSemantic && indexingMap.getNumResults() != rank)
       return op->emitOpError("expected operand rank (")
              << rank << ") to match the result rank of indexing_map #"
              << opOperand.getOperandNumber() << " ("
              << indexingMap.getNumResults() << ")";
-  }
 
+    // Try to verify broadcast/transpose semantic requested through explicit
```

Comment:
You dont need to do this here.  Add a `hasVerifier = 1` to the op definition in tablegen file. Then you would need to define a `LogicalResult MatmulOp::verify()` method. You can make this check in that method. AFAICS you should not need any change to the verifier of structured op semantics, but maybe I am missing something here?

Reply:
Since I was not sure what later part of the `verifyStructuredOpInterface` logic is required for MatmulOp, I thought of using it to trigger the verification of matmulOp with extended semantic and let it go through the later part of the logic.

However, I can duplicate those logic and completely separate the verification of MatmulOp. Are you asking for this?

Reply:
We don't need to verify everything in all places, as long as it's checked in one place, it's ok.

So here, just move the call to `verifyExtendedMatmulSemantic` to `Matmul::verify()` and we're done.

Reply:
I was thinking that Ops verifier will override the verifyStructuredOpInterface() but that is not the case. Thanks for the clarification, moved it to `Matmul::verify()`.